### PR TITLE
Dependencies API tweaks

### PR DESCRIPTION
**Pull request by lukaspawlik**
@spacedmonkey this is still WIP however, it addresses most meaningful feedback found on https://github.com/WordPress/gutenberg/pull/21244

### CODE REVIEW

File: lib/class-wp-rest-styles-controller.php [hunk 1]
```diff
@@ -39,11 +47,13 @@ public function get_url( $src, $ver, $handle ) {
 	 * @return array
 	 */
 	public function get_core_assets() {
-		$wp_styles = new WP_Styles();
-		wp_default_styles( $wp_styles );
-		$handles = wp_list_pluck( $wp_styles->registered, 'handle' );
-		$handles = array_values( $handles );
+		if ( null === $this->core_assets ) {
```
Comment (spacedmonkey) [inline lib/class-wp-rest-styles-controller.php:43]: Why cache has these?

Comment (lukaspawlik) [reply lib/class-wp-rest-styles-controller.php:43]: @spacedmonkey I noticed that this function is called twice during a single request and it's not harmful at all to cache it for the second execution. It's not particularly slow but basically no real need to run this code the second time.

File: lib/class-wp-rest-dependencies-controller.php [hunk 2]
```diff
@@ -98,6 +100,9 @@ public function get_items( $request ) {
 
 		if ( $handle ) {
 			foreach ( $filter as $dependency_handle ) {
+				if ( $handle === $dependency_handle ) {
```
Comment (spacedmonkey) [inline lib/class-wp-rest-dependencies-controller.php:110]: 👍

Comment (spacedmonkey) [reply lib/class-wp-rest-dependencies-controller.php:110]: Can we add a unit test for this?

**Comment by lukaspawlik** [comment]
@spacedmonkey I think some of these tests fail as they are executed out of main repo - All PHP stuff tests pass on my local without any issues.

File: phpunit/class-rest-block-types-controller-test.php [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+		parent::setUp();
+	}
+
+	/**
+	 * Tear down each test method.
+	 */
+	public function tearDown() {
+		parent::tearDown();
+	}
+
```
Comment (spacedmonkey) [inline phpunit/class-rest-block-types-controller-test.php]:
```suggestion

```

No needed.

File: phpunit/class-rest-block-types-controller-test.php [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	public static function wpSetUpBeforeClass( $factory ) {
+		self::$admin_id = $factory->user->create(
+			array(
+				'role' => 'administrator',
+			)
+		);
+		wp_register_script( 'core-assets-test-script', home_url( '/assets/test-script.js' ) );
+		wp_register_script( 'core-assets-test-editor-script', home_url( '/assets/test-editor-script.js' ) );
+		wp_register_style( 'core-assets-test-style', home_url( '/assets/test-style.css' ) );
+		wp_register_style( 'core-assets-test-editor-style', home_url( '/assets/test-editor-style.css' ) );
```
Comment (spacedmonkey) [inline phpunit/class-rest-block-types-controller-test.php:34]: One thing I would love to test here is weird inter dependencies. What happens when you have something that has dependencies of a dependencies. How does WordPress resolve that?

File: phpunit/class-rest-scripts-controller-test.php [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+		$this->assertCount( 3, $data );
+
+		$this->assertEquals( 'dependency3', $data[0]['handle'] );
+		$this->assertEquals( 'dependency4', $data[1]['handle'] );
+		$this->assertEquals( 'dependency5', $data[2]['handle'] );
+
+		for ( $i = 0; $i < 3; $i ++ ) {
+			$keys = array( 'src', 'url', 'args', 'ver', 'extra', 'textdomain', 'translations_path', 'deps', '_links' );
```
Comment (spacedmonkey) [inline phpunit/class-rest-scripts-controller-test.php]: Dont define key in loop.

File: phpunit/class-rest-scripts-controller-test.php [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+		wp_register_script( 'script-with-nested-deps', home_url( '/script-with-nested-deps.js' ), array( 'dependency5' ) );
+	}
+
+	/**
+	 * Tests tear down.
+	 */
+	public function tearDown() {
+		parent::tearDown();
+	}
+
```
Comment (spacedmonkey) [inline phpunit/class-rest-scripts-controller-test.php]:
```suggestion

```

File: phpunit/class-rest-scripts-controller-test.php (hunk 5 above)
Comment (lukaspawlik) [reply phpunit/class-rest-scripts-controller-test.php]: 🤦

File: lib/class-wp-rest-scripts-controller.php [hunk 7]
```diff
@@ -9,6 +9,14 @@
  * Class WP_REST_Scripts_Controller
  */
 class WP_REST_Scripts_Controller extends WP_REST_Dependencies_Controller {
+
+	/**
+	 * Core assets collection.
+	 *
+	 * @var null|array
+	 */
+	protected $core_assets;
```
Comment (spacedmonkey) [inline lib/class-wp-rest-scripts-controller.php]: Cant we just add this to the `WP_REST_Dependencies_Controller` class?

File: lib/class-wp-rest-styles-controller.php [hunk 8]
```diff
@@ -9,6 +9,14 @@
  * Class WP_REST_Styles_Controller
  */
 class WP_REST_Styles_Controller extends WP_REST_Dependencies_Controller {
+
+	/**
+	 * Core assets collection.
+	 *
+	 * @var null|array
+	 */
+	protected $core_assets;
```
Comment (spacedmonkey) [inline lib/class-wp-rest-styles-controller.php]: Cant we just add this to the `WP_REST_Dependencies_Controller` class?

File: phpunit/class-rest-block-types-controller-test.php (hunk 4 above)
Comment (lukaspawlik) [reply phpunit/class-rest-block-types-controller-test.php:34]: @spacedmonkey Okay. I will try to re-implement the approach from scripts/styles controllers test where there are nested deps.

Comment (spacedmonkey) [reply phpunit/class-rest-block-types-controller-test.php:34]: No, I think you are good with what you have. I would have a new test with really messed on dependencies. Like three levels, just to see what happens and how it breaks.

File: lib/class-wp-rest-styles-controller.php (hunk 8 above)
Comment (lukaspawlik) [reply lib/class-wp-rest-styles-controller.php]: @spacedmonkey Moved there.

File: lib/class-wp-rest-scripts-controller.php (hunk 7 above)
Comment (lukaspawlik) [reply lib/class-wp-rest-scripts-controller.php]: @spacedmonkey Moved there.

File: phpunit/class-rest-scripts-controller-test.php (hunk 5 above)
Comment (lukaspawlik) [reply phpunit/class-rest-scripts-controller-test.php]: Fixed in both places. Some leftover before I ran this code in a loop but still 🤦

File: lib/rest-api.php [hunk 9]
```diff
@@ -369,3 +369,49 @@ function gutenberg_term_search_handler( $search_handlers ) {
...
+	$links   = array();
+	$scripts = array( 'editor_script', 'script' );
+	foreach ( $scripts as $script ) {
+		if ( ! isset( $block_type->$script ) ) {
+			continue;
+		}
+		$expected_handle = $block_type->$script;
+		if ( wp_script_is( $expected_handle, 'registered' ) ) {
+			$links[ 'https://api.w.org/' . $script ] = array(
+				'href'       => rest_url( sprintf( '%s/%s/%s', 'wp/v2', 'scripts', $expected_handle ) ),
```
Comment (spacedmonkey) [inline lib/rest-api.php]:
```suggestion
				'href'       => rest_url( sprintf( '%s/%s/%s', '__experimental', 'scripts', $expected_handle ) ),
```

File: lib/rest-api.php [hunk 10]
```diff
@@ -369,3 +369,49 @@ function gutenberg_term_search_handler( $search_handlers ) {
...
+
+	$styles = array( 'editor_style', 'style' );
+	foreach ( $styles as $style ) {
+		if ( ! isset( $block_type->$style ) ) {
+			continue;
+		}
+		$expected_handle = $block_type->$style;
+		if ( wp_style_is( $expected_handle, 'registered' ) ) {
+			$links[ 'https://api.w.org/' . $style ] = array(
+				'href'       => rest_url( sprintf( '%s/%s/%s', 'wp/v2', 'styles', $expected_handle ) ),
```
Comment (spacedmonkey) [inline lib/rest-api.php]:
```suggestion
				'href'       => rest_url( sprintf( '%s/%s/%s', '__experimental', 'styles', $expected_handle ) ),
```

File: phpunit/class-rest-block-types-controller-test.php (hunk 4 above)
Comment (lukaspawlik) [reply phpunit/class-rest-block-types-controller-test.php:34]: @spacedmonkey Not sure if that test would apply here. Block types tests are only to test for proper `_links` entry to be returned like following example:

```
"wp:editor_script": [
            {
                "embeddable": true,
                "href": "http://localhost:8889/wp-json/wp/v2/scripts/core-assets-test"
            }
        ],
```

We don't seem to be returning here any dependencies or deeper relationships.

File: lib/class-wp-rest-dependencies-controller.php (hunk 2 above)
Comment (lukaspawlik) [reply lib/class-wp-rest-dependencies-controller.php:110]: @spacedmonkey I've added dedicated test for it, even it above solution was covered by `get_items` tests (like checks of number of dependencies returned etc).

File: phpunit/class-rest-styles-controller-test.php [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+		wp_register_style( 'style-with-deps', home_url( '/style-with-deps.css' ), array( 'dependency1', 'dependency2' ) );
+		wp_register_style( 'style-with-nested-deps', home_url( '/style-with-nested-deps.css' ), array( 'dependency5' ) );
+	}
+
+	/**
+	 * Tests tear down.
+	 */
+	public function tearDown() {
+		parent::tearDown();
+	}
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]:
```suggestion

```

File: phpunit/class-rest-styles-controller-test.php [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+		$this->assertArrayHasKey( 'extra', $properties );
+		$this->assertArrayHasKey( 'textdomain', $properties );
+		$this->assertArrayHasKey( 'translations_path', $properties );
+		$this->assertArrayHasKey( 'deps', $properties );
+	}
+
+	/**
+	 * Tests set up.
+	 */
+	public function setUp() {
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]: Nit pic, setup normally appears at the top of the class.

File: phpunit/class-rest-styles-controller-test.php [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+			array(
+				'role' => 'editor',
+			)
+		);
+		self::$author_id     = $factory->user->create(
+			array(
+				'role' => 'author',
+			)
+		);
+		self::$subscriber_id = $factory->user->create(
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]: We should use this user to test auth. Use a get_item / get_items call as a subscriber

File: phpunit/class-rest-styles-controller-test.php [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+		);
+		if ( is_multisite() ) {
+			update_site_option( 'site_admins', array( 'superadmin' ) );
+		}
+		self::$admin_id      = $factory->user->create(
+			array(
+				'role' => 'administrator',
+			)
+		);
+		self::$editor_id     = $factory->user->create(
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]: What is this used for ?

File: phpunit/class-rest-styles-controller-test.php [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+			array(
+				'role' => 'editor',
+			)
+		);
+		self::$author_id     = $factory->user->create(
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]: Can we test to see you can see block assets and their dependencies as an author.

File: phpunit/class-rest-styles-controller-test.php [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+		wp_set_current_user( self::$admin_id );
+		$request  = new WP_REST_Request( 'GET', '/__experimental/styles/' . self::$style_handle );
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+
+		$this->assertEquals( self::$style_handle, $data['handle'] );
+		$this->assertEquals( home_url( '/test.css' ), $data['src'] );
+		$this->assertEquals( home_url( '/test.css' ), $data['url'] );
+	}
+
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php:204]: Add tests for no user / subscriber / author to see that permissions checks work.

File: phpunit/class-rest-styles-controller-test.php (hunk 12 above)
Comment (lukaspawlik) [reply phpunit/class-rest-styles-controller-test.php]: 👍

File: phpunit/class-rest-styles-controller-test.php (hunk 13 above)
Comment (lukaspawlik) [reply phpunit/class-rest-styles-controller-test.php]: This has been added. Subscriber cannot view them according to tests.

File: phpunit/class-rest-styles-controller-test.php (hunk 14 above)
Comment (lukaspawlik) [reply phpunit/class-rest-styles-controller-test.php]: It's been used for tests. Unfortunately editors can't read scripts/styles. The same applies to authors.

File: phpunit/class-rest-styles-controller-test.php (hunk 15 above)
Comment (lukaspawlik) [reply phpunit/class-rest-styles-controller-test.php]: It's been used for tests. Unfortunately authors can't read scripts/styles. I think this should be addressed right?

File: phpunit/class-rest-styles-controller-test.php (hunk 16 above)
Comment (lukaspawlik) [reply phpunit/class-rest-styles-controller-test.php:204]: Added

File: phpunit/class-rest-styles-controller-test.php (hunk 15 above)
Comment (lukaspawlik) [reply phpunit/class-rest-styles-controller-test.php]: So tests are executed against not core assets and mentioned roles (authors, editors) can't see it. That's not a case for core assets. Shall we address this to ensure that authors, editors can also read scripts / styles? That doesn't seem to be risky at all.

File: phpunit/class-rest-scripts-controller-test.php [hunk 17]
```diff
@@ -0,0 +1,384 @@
...
+		$this->assertArrayHasKey( 'textdomain', $properties );
+		$this->assertArrayHasKey( 'translations_path', $properties );
+		$this->assertArrayHasKey( 'deps', $properties );
+	}
+
+	/**
+	 * Test single block script.
+	 */
+	public function test_get_item_block_script() {
+		foreach ( array( 0, self::$subscriber_id, self::$editor_id, self::$author_id, self::$admin_id, self::$superadmin_id ) as $user_id ) {
```
Comment (spacedmonkey) [inline phpunit/class-rest-scripts-controller-test.php]: I don't personally like this style of loop. Can we just use a dataprovider here and other places that this format is used? Using dataprovider really helps in debugging.

**Comment by spacedmonkey** [comment]
Barring this fix https://github.com/spacedmonkey/gutenberg/pull/1/files#r514452318, I think we are ready to merge @lukaspawlik 🎉

File: phpunit/class-rest-scripts-controller-test.php (hunk 17 above)
Comment (lukaspawlik) [reply phpunit/class-rest-scripts-controller-test.php]: @spacedmonkey I tried to use data providers and they were not delivering the right data - if they return primitives like integers then they work fine. When I tried to return class props they were always set to null. Any idea what I did incorrectly?

File: phpunit/class-rest-styles-controller-test.php [hunk 18]
```diff
@@ -0,0 +1,461 @@
...
+		$this->assertNotContains( 'style-with-nested-deps', $handles );
+		$this->assertContains( 'dependency3', $handles );
+		$this->assertContains( 'dependency4', $handles );
+	}
+
+	/**
+	 * Test single style.
+	 */
+	public function test_get_item() {
+		foreach ( array( self::$admin_id, self::$superadmin_id ) as $user_id ) {
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]: Same here.

File: phpunit/class-rest-styles-controller-test.php [hunk 19]
```diff
@@ -0,0 +1,461 @@
...
+		$data     = $response->get_data();
+		$this->assertEquals( 'view', $data['endpoints'][0]['args']['context']['default'] );
+		$this->assertEquals( array( 'view', 'embed', 'edit' ), $data['endpoints'][0]['args']['context']['enum'] );
+	}
+
+	/**
+	 * Test multiple styles.
+	 */
+	public function test_get_items() {
+		foreach ( array( self::$admin_id, self::$superadmin_id ) as $user_id ) {
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]: Same here.

File: phpunit/class-rest-scripts-controller-test.php [hunk 20]
```diff
@@ -0,0 +1,461 @@
...
+
+	/**
+	 * Tear down tests after entire test class is done.
+	 */
+	public static function wpTearDownAfterClass() {
+		self::delete_user( self::$superadmin_id );
+		self::delete_user( self::$admin_id );
+		self::delete_user( self::$editor_id );
+		self::delete_user( self::$author_id );
+		self::delete_user( self::$subscriber_id );
```
Comment (spacedmonkey) [inline phpunit/class-rest-scripts-controller-test.php]:
```suggestion
	foreach( self::$users_map as $key => $user_id ){
            self::delete_user( $user_id );
        }
```

File: phpunit/class-rest-styles-controller-test.php [hunk 21]
```diff
@@ -0,0 +1,461 @@
...
+
+	/**
+	 * Tear down tests after entire test class is done.
+	 */
+	public static function wpTearDownAfterClass() {
+		self::delete_user( self::$superadmin_id );
+		self::delete_user( self::$admin_id );
+		self::delete_user( self::$editor_id );
+		self::delete_user( self::$author_id );
+		self::delete_user( self::$subscriber_id );
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]:
```suggestion
	foreach( self::$users_map as $key => $user_id ){
            self::delete_user( $user_id );
        }
```

File: phpunit/class-rest-styles-controller-test.php [hunk 22]
```diff
@@ -0,0 +1,461 @@
...
+		self::$users_map['superadmin'] = self::$superadmin_id;
+		if ( is_multisite() ) {
+			update_site_option( 'site_admins', array( 'superadmin' ) );
+		}
+		self::$admin_id                = $factory->user->create(
+			array(
+				'role' => 'administrator',
+			)
+		);
+		self::$users_map['admin']      = self::$admin_id;
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]:
```suggestion
		self::$users_map['admin']      = $factory->user->create(
			array(
				'role' => 'administrator',
			)
		);
```

File: phpunit/class-rest-styles-controller-test.php [hunk 23]
```diff
@@ -0,0 +1,461 @@
...
+class REST_Styles_Controller_Test extends WP_Test_REST_Controller_Testcase {
+	/**
+	 * @var int
+	 */
+	protected static $superadmin_id;
+
+	/**
+	 * @var int
+	 */
+	protected static $admin_id;
```
Comment (spacedmonkey) [inline phpunit/class-rest-styles-controller-test.php]: We can get rid of these class properties as well.

**Comment by lukaspawlik** [comment]
@spacedmonkey ready for your review (another one 😄 )

**Review by spacedmonkey — APPROVED** [review]
👍 Looking good.

**Pull request merged**